dm/syncer/ddl.go

```diff
@@ -192,6 +192,45 @@ func (s *Syncer) clearOnlineDDL(tctx *tcontext.Context, targetTable *filter.Tabl
 	return nil
 }
 
+func (s *Syncer) adjustTableCollation(ddlInfo *ddlInfo) error {
+	if createStmt, ok := ddlInfo.originStmt.(*ast.CreateTableStmt); ok {
+		// create table like old table has no table_options
+		// See https://dev.mysql.com/doc/refman/5.7/en/create-table.html
+		if createStmt.ReferTable != nil {
+			return nil
+		}
+		for _, tableOption := range createStmt.Options {
+			// already have 'Collation'
+			if tableOption.Tp == ast.TableOptionCollate {
+				return nil
+			}
+		}
+		sourceTable := ddlInfo.sourceTables[0]
+		collationInfo, err := utils.GetTableCollation(s.tctx.Ctx, s.fromDB.BaseDB.DB, sourceTable.Schema, sourceTable.Name)
+		if err != nil {
+			s.tctx.L().Error("get source table collation info failed.", zap.Stringer("table", sourceTable), zap.Error(err))
+			return err
+		}
+		if collationInfo == "" {
+			return nil
+		}
+		// add collation
+		if createStmt.Options == nil {
+			createStmt.Options = make([]*ast.TableOption, 1)
+		}
+		createStmt.Options = append(createStmt.Options, &ast.TableOption{Tp: ast.TableOptionCollate, StrValue: collationInfo})
+		routedDDL, err := parserpkg.RenameDDLTable(createStmt, ddlInfo.targetTables)
```

Comment:
we can move invocation of `adjustTableCollation` inside `genDDLInfo` to avoid this

Reply:
ok

Reply:
There's nothing to batch. If the query event is CREATE TABLE, after processing there should be at most one DDL to handle 🤔

Reply:
ok

Reply:
i think lance suggestion is better